### PR TITLE
Fixed liquidity management

### PR DESCRIPTION
**Pull request by Vovchyk**

### CODE REVIEW

File: providers/liquidity.go [hunk 1]
```diff
@@ -26,25 +27,24 @@ import (
 )
 
 type LiquidityProvider interface {
-	GetQuote(types.Quote, uint64, uint64) *types.Quote
 	Address() string
-	SignQuote(hash []byte, depositAddr string, reqLiq *big.Int) ([]byte, error)
+	GetQuote(*types.Quote, uint64, *types.Wei) (*types.Quote, error)
+	SignQuote(hash []byte, depositAddr string, reqLiq *types.Wei) ([]byte, error)
 	SignTx(common.Address, *gethTypes.Transaction) (*gethTypes.Transaction, error)
-	SetLiquidity(value *big.Int)
```
Comment (Vovchyk) [inline providers/liquidity.go:33]: no more need to keep liquidity amount in memory, it's now being fetched at runtime

File: providers/liquidity.go [hunk 2]
```diff
@@ -79,11 +75,13 @@ func NewLocalProvider(config ProviderConfig, repository RetainedQuotesRepository
...
 		}
-		defer f.Close()
+		defer func(f *os.File) {
+			_ = f.Close()
+		}(f)
 	}
 
 	ks := keystore.NewKeyStore(config.Keydir, keystore.StandardScryptN, keystore.StandardScryptP)
-	acc, err := retreiveOrCreateAccount(ks, config.AccountNum, f)
+	acc, err := retrieveOrCreateAccount(ks, config.AccountNum, f)
```
Comment (Vovchyk) [inline providers/liquidity.go:84]: typo

File: providers/liquidity.go [hunk 3]
```diff
@@ -162,21 +136,34 @@ func (lp *LocalProvider) SignQuote(hash []byte, depositAddr string, reqLiq *big.
 	}
 	signB[len(signB)-1] += 27 // v must be 27 or 28
 
-	if rq == nil {
+	lp.mu.Lock()
```
Comment (Vovchyk) [inline providers/liquidity.go:139]: lock to prevent race conditions in liquidity acquisition

File: providers/liquidity.go [hunk 4]
```diff
@@ -57,15 +57,11 @@ type ProviderConfig struct {
 	Confirmations  map[int]uint16
 	TimeForDeposit uint32
 	CallTime       uint32
-	CallFee        uint64
-	PenaltyFee     uint64
+	CallFee        *types.Wei
+	PenaltyFee     *types.Wei
 }
 
-type InMemRetainedQuotesRepository struct {
```
Comment (Vovchyk) [inline providers/liquidity.go:64]: moved this to the test class

File: types/quote.go [hunk 5]
```diff
@@ -7,13 +7,13 @@ type Quote struct {
...
-	CallFee            uint64 `json:"callFee" db:"call_fee"`
-	PenaltyFee         uint64 `json:"penaltyFee" db:"penalty_fee"`
+	CallFee            *Wei   `json:"callFee" db:"call_fee"`
+	PenaltyFee         *Wei   `json:"penaltyFee" db:"penalty_fee"`
 	ContractAddr       string `json:"contractAddr" db:"contract_addr"`
 	Data               string `json:"data" db:"data"`
 	GasLimit           uint32 `json:"gasLimit" db:"gas_limit"`
 	Nonce              int64  `json:"nonce" db:"nonce"`
-	Value              uint64 `json:"value" db:"value"`
+	Value              *Wei   `json:"value" db:"value"`
```
Comment (Vovchyk) [inline types/quote.go:16]: use `Wei`'s (alias / extension for `big.Int`) for amounts

File: types/wei.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
+package types
+
+import (
+	"database/sql/driver"
+	"errors"
+	"math/big"
+)
+
+type Wei big.Int
```
Comment (Vovchyk) [inline types/wei.go:9]: new type that can handle big amounts

File: types/wei.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+func (w *Wei) ToSatoshi() *big.Float {
+	return new(big.Float).Quo(new(big.Float).SetInt(w.AsBigInt()), new(big.Float).SetInt(bTenPowTen))
+}
+
+func (w *Wei) String() string {
+	return w.AsBigInt().String()
+}
+
+func (w *Wei) Value() (driver.Value, error) {
```
Comment (Vovchyk) [inline types/wei.go:71]: used for marshalling in db

File: types/wei.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	if w == nil {
+		return "", errors.New("cannot retrieve value from <nil>")
+	}
+	return w.AsBigInt().String(), nil
+}
+
+func (w *Wei) Scan(src interface{}) error {
```
Comment (Vovchyk) [inline types/wei.go:78]: used for unmarshalling in db

**Review by josedahlquist — APPROVED** [review]

**Pull request merged**